aten/src/ATen/native/mps/operations/FourierTransform.mm

```diff
@@ -0,0 +1,205 @@
+
+  @autoreleasepool {
+    string key = key_prefix + getTensorsStringKey({input}) + getMPSTypeString(input);
+
+    auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(key, [&](auto mpsGraph, auto newCachedGraph) {
+      MPSGraphTensor* inputTensor = mpsGraphRankedPlaceHolder(mpsGraph, getMPSDataType(input), getMPSShape(input));
+      MPSGraphTensor* outputTensor = createFFTGraph(mpsGraph, inputTensor, signal_ndim, normalized, fftType, forward);
+      newCachedGraph->inputTensor = inputTensor;
+      newCachedGraph->outputTensor = outputTensor;
+    });
+
+    // Use utility functions to determine the shape of the output tensor
+    auto output_shape = getMPSShape(input);
+    int64_t output_height = output_shape[0].integerValue;
+    int64_t output_width = output_shape[1].integerValue;
+
+    MPSNDArrayDescriptor* sizeDesc = [MPSNDArrayDescriptor descriptorWithDataType:MPSDataTypeInt32 shape:@[ @(2) ]];
+    MPSNDArray* sizeNDArray = [[[MPSNDArray alloc] initWithDevice:stream->device() descriptor:sizeDesc] autorelease];
+    [sizeNDArray writeBytes:(int32_t[]){(int32_t)output_height, (int32_t)output_width} strideBytes:nil];
+    MPSGraphTensorData* sizeTensorData = [[[MPSGraphTensorData alloc] initWithMPSNDArray:sizeNDArray] autorelease];
+
+    Placeholder inputPlaceholder = Placeholder(cachedGraph->inputTensor, input);
+    Placeholder outputPlaceholder =
+        Placeholder(cachedGraph->outputTensor, out.has_storage() ? out : output, nil, false);
+
+    NSDictionary<MPSGraphTensor*, MPSGraphTensorData*>* feeds = @{
+      inputPlaceholder.getMPSGraphTensor() : inputPlaceholder.getMPSGraphTensorData(),
+      cachedGraph->outputSizeTensor : sizeTensorData,
+    };
+    NSDictionary<MPSGraphTensor*, MPSGraphTensorData*>* results =
+        @{outputPlaceholder.getMPSGraphTensor() : outputPlaceholder.getMPSGraphTensorData()};
+    runMPSGraph(stream, cachedGraph->graph(), feeds, results);
+
+    if (out.has_storage()) {
+      output.copy_(out);
+    }
+  }
+}
+
+} // namespace mps
+
+static bool check_mps_compatibility() {
+  static const bool is_macOS_14_0_or_newer = is_macos_13_or_newer(MacOSVersion::MACOS_VER_14_0_PLUS);
+  if (!is_macOS_14_0_or_newer) {
+    TORCH_WARN_ONCE("MPS: FFT operations are only supported natively starting from macOS 14.0. ",
+                    "Falling back on CPU. This may have performance implications.");
+    return false;
+  }
+  return true;
+}
+
+// Real-to-Hermitean FFT
+Tensor _fft_r2c_mps(const Tensor& input, int64_t signal_ndim, bool normalized, IntArrayRef signal_sizes) {
+  return runFFTGraph(input, {}, false, "fft_r2c_mps", mps::FFTType::R2C, signal_ndim, normalized, true);
+}
+
+// Real-to-Hermitean FFT, writing result to the provided output tensor
+Tensor& _fft_r2c_mps_out(const Tensor& input,
+                         Tensor& out,
+                         int64_t signal_ndim,
+                         bool normalized,
+                         IntArrayRef signal_sizes) {
```

Comment:
Sorry, but where did you get this function signature from?

Reply:
Hey there @beveradb, Just wondering, I saw for your projects you converted to using MLX, is there still interest in working on this? I'd be willing to sponsor a bounty to get this over the line if the original motivation has gone :-)

Reply:
> Hey there @beveradb, Just wondering, I saw for your projects you converted to using MLX, is there still interest in working on this? I'd be willing to sponsor a bounty to get this over the line if the original motivation has gone :-)

Put me in for contributing towards the bounty as well as long as my portion is reasonable.

Reply:
I'm not actually sure how to best do the bounty, the platform I used to use seems to have disappeared and taken everyones money, but I'd be willing to contribute 200 USD towards seeing the FFT ops land in pytorch, I think they're the last ops needed for GPU accelerated TTS/Audio manipulation

Reply:
> I'm not actually sure how to best do the bounty, the platform I used to use seems to have disappeared and taken everyones money, but I'd be willing to contribute 200 USD towards seeing the FFT ops land in pytorch, I think they're the last ops needed for GPU accelerated TTS/Audio manipulation

My landlord would shreak if I contributed more than $20...sorry, but hope it'll help.

Reply:
Appreciate the offer guys, the motivation is still there but I've had to prioritize all of my after-work hours towards other paid work recently due to unplanned family/personal life events.
If nobody else beats me to it I will eventually get back to working on this again when I have more free time, but it's honestly mostly a mental blocker of me feeling out of my depth with it which is the bigger issue; if it was easy I'd have done it already.
If you want it to be implemented sooner, find someone who's actually an experienced Objective-C / C++ developer, preferably with PyTorch and/or macOS development experience, as I'm just fumbling and guessing / doing my best with limited understanding!

Reply:
That's fair enough! Sorry to hear that @beveradb hope everything works out okay. Yeah it's not a trivial implementation, I'd work on it myself but I have almost zero objective-c experience and minimal pytorch C++ code experience. I really appreciate your efforts, your life and family come first of course. Hope it all works out :-)

Reply:
Hey @malfet sorry for the @, just wondering if you'd know anyone who'd be interested in completing this for a bounty? I know it's not a lot compared to the amount of experience required to work on it, but I'm a single person, not a big company or anything

Reply:
@agnosticlines FFT should be available in nightlies (committed as https://github.com/pytorch/pytorch/pull/119670 ), if something still does not work, please file an issue